### PR TITLE
Editor: Add collapsing carousel

### LINKED ISSUE #9641: Workspace UX changes: Allow collapsing page carousel
## Feature Description
See [Figma](https://www.figma.com/file/dc9xrVlRhQq4KTE9WQzU1R/_Editor---Sprint-Feature-Level-Designs_?node-id=10638%3A453154)

- [ ] Add button for collapsing page carousel that shows the total number of pages + the selected page number
- [ ] Allow collapsing page carousel, add animation for sliding up/down
- [ ] The carousel will display as many pages as fit (as it currently does)
- [ ] Keep in mind the thumbnails generation when implementing to ensure it still works as expected.

Notes: 
- the default state of the page carousel is expanded, no need to store the state.
- increasing page size to use the new room will be done separately

## Additional Context

Expanded vs collapsed states:
Screenshot 2021-11-05 at 13 22 58

### PR DESCRIPTION
**Pull request by barklund**
## Context

This adds a collapsing carousel as per Figma designs.

## To-do

- [x] Tests

## User-facing changes

| State | Video |
|-|-|
| Wide page | carousel-collapse-wide-fast |
| Narrow page with many thumbnails | carousel-collapse-narrow-multipage-fast |
| Narrow page with a single thumbnail | carousel-collapse-narrow-singlepage-fast |

## Testing Instructions

This PR can be tested by following these steps:

1. Observe there is a new bar above the page carousel with the current page number and the total page count
2. Click this bar to collapse the page carousel
3. Observe the collapse state with the page count bar perfectly aligning in the middle of the buttons to either side and that the canvas above has expanded to use the extra space
4. Click the bar again to expand the page carousel
5. Observe that everything works as expected and that the canvas has shrunk to make room for the carousel.

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #9641

### CODE REVIEW

File: packages/story-editor/src/components/footer/carousel/karma/carousel.karma.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import { Fixture } from '../../../../karma';
+
+describe('Carousel', () => {
```
Comment (miina) [inline packages/story-editor/src/components/footer/carousel/karma/carousel.karma.js]: We also have a `carousel.karma.js` file in the canvas folder (search for `'Carousel integration'` to find it) -- should we perhaps use one file for carousel only?

File: packages/story-editor/src/components/footer/carousel/carouselLayout.js [hunk 2]
```diff
@@ -54,24 +98,53 @@ const Area = styled.div`
...
-        <CarouselScrollBack />
-      </Area>
-      <Area area="carousel">
-        <CarouselList />
-      </Area>
-      <Area area="next-navigation">
-        <CarouselScrollForward />
-      </Area>
-    </Wrapper>
+    <CSSTransition in={isOpenOrOpening} classNames="carousel" timeout={700}>
```
Comment (miina) [inline packages/story-editor/src/components/footer/carousel/carouselLayout.js]: Use the constant instead of 700 here?

**Comment by miina** [comment]
From testing:

- In some cases, it looks like the left navigation arrow can be covered by the buttons. It coincidentally happened exactly with the width of the browser that I was using when checking out the PR.
Screenshot 2021-11-17 at 11 44 42

- Previously, it was possible to navigate between the Pages using arrow keys, it doesn't seem to be possible anymore, so once tabbed into a page preview, there's no way to focus on the next page using keyboard.

- I wonder if the animation should be a bit faster?

**Review by miina — APPROVED** [review]
Looks great!

Added a few nits + some comments from testing that shouldn't be hard to fix.

**Comment by swissspidy** [comment]
> I wonder if the animation should be a bit faster?

💯 

The animation in the GIF above seems very slow.

**Comment by barklund** [comment]
> * In some cases, it looks like the left navigation arrow can be covered by the buttons. It coincidentally happened exactly with the width of the browser that I was using when checking out the PR.

As discussed in slack, this is a very strange issue that I cannot recreate. I have removed some superfluous margins, but not sure that'll fix it on your machine. Let's see if it's still a reproducible issue after merge.

> * Previously, it was possible to navigate between the Pages using arrow keys, it doesn't seem to be possible anymore, so once tabbed into a page preview, there's no way to focus on the next page using keyboard.

Fixed - good find!

> * I wonder if the animation should be a bit faster?

Increase to 300ms

File: packages/story-editor/src/components/footer/carousel/karma/carousel.karma.js (hunk 1 above)
Comment (barklund) [reply packages/story-editor/src/components/footer/carousel/karma/carousel.karma.js]: Rearranged in 0f873a462c. I was a bit surprised that we didn't have any tests actually...

File: packages/story-editor/src/components/footer/carousel/carouselLayout.js (hunk 2 above)
Comment (barklund) [reply packages/story-editor/src/components/footer/carousel/carouselLayout.js]: Done in 078eda9b34. Also increased the speed to 300ms.

**Comment by barklund** [comment]
> The animation in the GIF above seems very slow.

GIFs are now updated after a speed increase to 300ms.

File: packages/story-editor/src/components/footer/carousel/useCarouselScroll.js [hunk 3]
```diff
@@ -97,9 +107,19 @@ function useCarouselScroll({
     [pageThumbWidth, pageThumbMargin]
   );
 
+  // Is this the first scroll (which will be instant rather than animated)?
+  const firstScroll = useRef(true);
+
+  // If the carousel drawer is collapsed, reset first scroll to true
+  useEffect(() => {
+    if (carouselState == CAROUSEL_STATE.CLOSED) {
```
Comment (miina) [inline packages/story-editor/src/components/footer/carousel/useCarouselScroll.js]:
```suggestion
    if (carouselState === CAROUSEL_STATE.CLOSED) {
```

File: packages/story-editor/src/components/footer/carousel/useCarouselScroll.js [hunk 4]
```diff
@@ -97,9 +107,19 @@ function useCarouselScroll({
...
+      firstScroll.current = true;
+    }
+  });
```
Comment (miina) [inline packages/story-editor/src/components/footer/carousel/useCarouselScroll.js]: Deps?

File: packages/story-editor/src/components/footer/carousel/karma/carouselNavigation.karma.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    fixture = new Fixture();
+    fixture.setPages([
+      { id: 'page1', backgroundColor: createSolid(255, 255, 255) },
+      { id: 'page2', backgroundColor: createSolid(255, 0, 0) },
+      { id: 'page3', backgroundColor: createSolid(0, 255, 0) },
+      { id: 'page4', backgroundColor: createSolid(0, 0, 255) },
+    ]);
+    await fixture.render();
+
+    const insertElement = await fixture.renderHook(() => useInsertElement());
```
Comment (miina) [inline packages/story-editor/src/components/footer/carousel/karma/carouselNavigation.karma.js]: Could we do this by actually clicking on a text element instead? If I'm not mistaken we even had an issue somewhere to replace using hooks for element insertion by using clicking -- to make it more close to real usage.

File: packages/story-editor/src/components/footer/carousel/karma/carouselDrawer.karma.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+      // Both scroll buttons exist
+      expect(fixture.editor.footer.carousel.next).toBeTruthy();
+      expect(fixture.editor.footer.carousel.previous).toBeTruthy();
+
+      await fixture.snapshot();
+    });
+
+    it('should collapse and expand', async () => {
+      await fixture.events.click(fixture.editor.footer.carousel.toggle);
+      await fixture.events.sleep(800);
```
Comment (miina) [inline packages/story-editor/src/components/footer/carousel/karma/carouselDrawer.karma.js]: Nit: Was this based on the animation time? If yes, probably could be reduced a bit now?

**Comment by miina** [comment]
I'm seeing some "jumpiness" or slight lag while expanding the carousel with 30 pages.
Is this related to the thumbnail images generation topic mentioned in Slack? Should we merge both the Page Size adjustments and the upcoming fix for the thumb images generation into this PR and only then merge the changes all together?

File: packages/story-editor/src/components/footer/carousel/useCarouselScroll.js (hunk 3 above)
Comment (barklund) [reply packages/story-editor/src/components/footer/carousel/useCarouselScroll.js]: Completely missed this review! Fixed in fb365f14c9.

File: packages/story-editor/src/components/footer/carousel/useCarouselScroll.js (hunk 4 above)
Comment (barklund) [reply packages/story-editor/src/components/footer/carousel/useCarouselScroll.js]: Fixed in fb365f14c9 .

Comment (barklund) [reply packages/story-editor/src/components/footer/carousel/useCarouselScroll.js:118]: Fixed in fb365f14c9.

File: packages/story-editor/src/components/footer/carousel/karma/carouselNavigation.karma.js (hunk 5 above)
Comment (barklund) [reply packages/story-editor/src/components/footer/carousel/karma/carouselNavigation.karma.js]: Fixed in acf6fc10be.

File: packages/story-editor/src/components/footer/carousel/karma/carouselDrawer.karma.js (hunk 6 above)
Comment (barklund) [reply packages/story-editor/src/components/footer/carousel/karma/carouselDrawer.karma.js]: Fixed in acf6fc10be.

**Pull request merged**